File: packages/ssl-pinning/android/src/main/java/io/capkit/sslpinning/SSLPinningPlugin.kt
package io.capkit.sslpinning

import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import io.capkit.sslpinning.config.SSLPinningConfig
import io.capkit.sslpinning.error.SSLPinningError
import io.capkit.sslpinning.error.SSLPinningErrorMessages
import io.capkit.sslpinning.logger.SSLPinningLogger
import io.capkit.sslpinning.model.SSLPinningResultModel

/**
 * Capacitor bridge for the SSLPinning plugin (Android).
 *
 * Responsibilities:
 * - Parse JavaScript input
 * - Call the native implementation
 * - Resolve or reject PluginCall
 * - Map native errors to JS-facing error codes
 */
@CapacitorPlugin(
  name = "SSLPinning",
  permissions = [
    Permission(
      alias = "network",
      strings = [android.Manifest.permission.INTERNET],
    ),
  ],
)
class SSLPinningPlugin : Plugin() {
  // -----------------------------------------------------------------------------
  // Properties
  // -----------------------------------------------------------------------------

  /**
   * Immutable plugin configuration read from capacitor.config.ts.
   * * CONTRACT:
   * - Initialized exactly once in `load()`.
   * - Treated as read-only afterwards.
   */
  private lateinit var config: SSLPinningConfig

  /**
   * Native implementation layer containing core Android logic.
   *
   * CONTRACT:
   * - Owned by the Plugin layer.
   * - MUST NOT access PluginCall or Capacitor bridge APIs directly.
   */
  private lateinit var implementation: SSLPinningImpl

  // -----------------------------------------------------------------------------
  // Companion Object
  // -----------------------------------------------------------------------------

  private companion object {
    /**
     * Account type identifier for internal plugin identification.
     */
    const val ACCOUNT_TYPE = "io.capkit.sslpinning"

    /**
     * Human-readable account name for the plugin.
     */
    const val ACCOUNT_NAME = "SSLPinning"
  }

  // -----------------------------------------------------------------------------
  // Lifecycle
  // -----------------------------------------------------------------------------

  /**
   * Called once when the plugin is loaded by the Capacitor bridge.
   *
   * This method initializes the configuration container and the native
   * implementation layer, ensuring all dependencies are injected.
   */
  override fun load() {
    super.load()

    config = SSLPinningConfig(this)
    implementation = SSLPinningImpl(context)
    implementation.updateConfig(config)

    SSLPinningLogger.debug("Plugin loaded. Version: ", BuildConfig.PLUGIN_VERSION)
  }

  // -----------------------------------------------------------------------------
  // Error Mapping
  // -----------------------------------------------------------------------------

  /**
   * Rejects the call with a message and a standardized error code.
   * Ensure consistency with the JS SSLPinningErrorCode enum.
   */
  private fun reject(
    call: PluginCall,
    error: SSLPinningError,
  ) {
    val code =
      when (error) {
        is SSLPinningError.Unavailable -> "UNAVAILABLE"
        is SSLPinningError.Cancelled -> "CANCELLED"
        is SSLPinningError.PermissionDenied -> "PERMISSION_DENIED"
        is SSLPinningError.InitFailed -> "INIT_FAILED"
        is SSLPinningError.InvalidInput -> "INVALID_INPUT"
        is SSLPinningError.UnknownType -> "UNKNOWN_TYPE"
        is SSLPinningError.NotFound -> "NOT_FOUND"
        is SSLPinningError.Conflict -> "CONFLICT"
        is SSLPinningError.Timeout -> "TIMEOUT"
        is SSLPinningError.NoPinningConfig -> "NO_PINNING_CONFIG"
        is SSLPinningError.CertNotFound -> "CERT_NOT_FOUND"
        is SSLPinningError.TrustEvaluationFailed -> "TRUST_EVALUATION_FAILED"
        is SSLPinningError.NetworkError -> "NETWORK_ERROR"
      }

    // Always use the message from the SSLPinningError instance
    val message = error.message ?: "Unknown native error"
    call.reject(message, code)
  }

  // -----------------------------------------------------------------------------
  // SSL Pinning (single fingerprint)
  // -----------------------------------------------------------------------------

  /**
   * Validates the SSL certificate of a HTTPS endpoint
   * using a single fingerprint.
   */
  @PluginMethod
  fun checkCertificate(call: PluginCall) {
    val url: String? = call.getString("url")
    val fingerprint: String? = call.getString("fingerprint")

    if (url.isNullOrBlank()) {
      call.reject(SSLPinningErrorMessages.URL_REQUIRED, "INVALID_INPUT")
      return
    }

    try {
      execute {
        try {
          val result: SSLPinningResultModel =
            implementation.checkCertificate(
              urlString = url ?: "",
              fingerprintFromArgs = fingerprint,
            )

          val jsResult = JSObject()
          jsResult.put("actualFingerprint", result.actualFingerprint)
          jsResult.put("fingerprintMatched", result.fingerprintMatched)
          jsResult.put("matchedFingerprint", result.matchedFingerprint)
          jsResult.put("excludedDomain", result.excludedDomain)
          jsResult.put("mode", result.mode)
          jsResult.put("error", result.error)
          jsResult.put("errorCode", result.errorCode)

          call.resolve(jsResult)
        } catch (error: SSLPinningError) {
          reject(call, error)
        } catch (error: Exception) {
          call.reject(
            error.message ?: SSLPinningErrorMessages.INTERNAL_ERROR,
            "INIT_FAILED",
          )
        }
      }
    } catch (error: Exception) {
      call.reject(
        error.message ?: SSLPinningErrorMessages.INTERNAL_ERROR,
        "INIT_FAILED",
      )
    }
  }

  // -----------------------------------------------------------------------------
  // SSL Pinning (multiple fingerprints)
  // -----------------------------------------------------------------------------

  /**
   * Validates the SSL certificate of a HTTPS endpoint
   * using multiple allowed fingerprints.
   */
  @PluginMethod
  fun checkCertificates(call: PluginCall) {
    val url: String? = call.getString("url")

    val jsArray: JSArray? = call.getArray("fingerprints")

    // Parsing JSArray to a clean Kotlin List
    val fingerprints: List<String>? =
      if (jsArray != null && jsArray.length() > 0) {
        val list = ArrayList<String>()
        for (i in 0 until jsArray.length()) {
          val value = jsArray.getString(i)
          if (!value.isNullOrBlank()) {
            list.add(value)
          }
        }
        if (list.isNotEmpty()) list else null
      } else {
        null
      }

    if (url.isNullOrBlank()) {
      call.reject(SSLPinningErrorMessages.URL_REQUIRED, "INVALID_INPUT")
      return
    }

    try {
      execute {
        try {
          val result: SSLPinningResultModel =
            implementation.checkCertificates(
              urlString = url ?: "",
              fingerprintsFromArgs = fingerprints,
            )

          val jsResult = JSObject()
          jsResult.put("actualFingerprint", result.actualFingerprint)
          jsResult.put("fingerprintMatched", result.fingerprintMatched)
          jsResult.put("matchedFingerprint", result.matchedFingerprint)
          jsResult.put("excludedDomain", result.excludedDomain)
          jsResult.put("mode", result.mode)
          jsResult.put("error", result.error)
          jsResult.put("errorCode", result.errorCode)

          call.resolve(jsResult)
        } catch (error: SSLPinningError) {
          reject(call, error)
        } catch (error: Exception) {
          call.reject(
            error.message ?: SSLPinningErrorMessages.INTERNAL_ERROR,
            "INIT_FAILED",
          )
        }
      }
    } catch (error: Exception) {
      call.reject(
        error.message ?: SSLPinningErrorMessages.INTERNAL_ERROR,
        "INIT_FAILED",
      )
    }
  }

  // -----------------------------------------------------------------------------
  // Version
  // -----------------------------------------------------------------------------

  /**
   * Returns the native plugin version.
   *
   * NOTE:
   * - This method is guaranteed not to fail
   * - Therefore it does NOT use TestError
   * - Version is injected at build time from package.json
   */
  @PluginMethod
  fun getPluginVersion(call: PluginCall) {
    val ret = JSObject()
    ret.put("version", BuildConfig.PLUGIN_VERSION)
    call.resolve(ret)
  }
}

File: packages/ssl-pinning/ios/Sources/SSLPinningPlugin/SSLPinningPlugin.swift
import Foundation
import Capacitor

/**
 Capacitor bridge for the SSLPinning plugin.

 Responsibilities:
 - Parse JavaScript input
 - Call the native implementation
 - Resolve or reject CAPPluginCall
 - Map native errors to JS-facing error codes
 */
@objc(SSLPinningPlugin)
public final class SSLPinningPlugin: CAPPlugin, CAPBridgedPlugin {

    // MARK: - Plugin metadata

    /// The unique identifier for the plugin.
    public let identifier = "SSLPinningPlugin"

    /// The name used to reference this plugin in JavaScript.
    public let jsName = "SSLPinning"

    /**
     * A list of methods exposed by this plugin. These methods can be called from the JavaScript side.
     * - `checkCertificate`: Validates an SSL certificate for a given URL.
     * - `checkCertificates`: Validates multiple SSL certificates for given URLs.
     * - `getPluginVersion`: Retrieves the current version of the plugin.
     */
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkCertificate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkCertificates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    // MARK: - Properties

    /// Native implementation containing platform-specific logic.
    private let implementation = SSLPinningImpl()

    /// Configuration instance
    private var config: SSLPinningConfig?

    // MARK: - Lifecycle

    /**
     Plugin lifecycle entry point.

     Called once when the plugin is loaded by the Capacitor bridge.
     This is the correct place to:
     - read static configuration
     - initialize native resources
     - configure the implementation
     */
    override public func load() {
        // Initialize SSLPinningConfig with the correct type
        let cfg = SSLPinningConfig(plugin: self)
        self.config = cfg
        implementation.applyConfig(cfg)

        // Log if verbose logging is enabled
        SSLPinningLogger.debug("Plugin loaded. Version: ", PluginVersion.number)
    }

    // MARK: - Error Mapping

    /**
     * Rejects the call using standardized error codes from the native SSLPinningError enum.
     */
    private func reject(
        _ call: CAPPluginCall,
        error: SSLPinningError
    ) {
        // Use the centralized errorCode and message defined in SSLPinningError.swift
        call.reject(error.message, error.errorCode)
    }

    private func handleError(_ call: CAPPluginCall, _ error: Error) {
        if let SSLPinningError = error as? SSLPinningError {
            call.reject(SSLPinningError.message, SSLPinningError.errorCode)
        } else {
            reject(call, error: .initFailed(error.localizedDescription))
        }
    }

    // MARK: - SSL Pinning (single fingerprint)

    /**
     Validates the SSL certificate of a HTTPS endpoint
     using a single fingerprint.
     */
    @objc func checkCertificate(_ call: CAPPluginCall) {
        let url = call.getString("url", "")
        let fingerprintValue = call.getString("fingerprint")
        let fingerprint =
            fingerprintValue?.isEmpty == false
            ? fingerprintValue
            : nil

        guard !url.isEmpty else {
            call.reject(
                SSLPinningErrorMessages.urlRequired,
                "INVALID_INPUT"
            )
            return
        }

        Task {
            do {
                let result =
                    try await implementation.checkCertificate(
                        urlString: url,
                        fingerprintFromArgs: fingerprint
                    )

                // Converting Swift Dictionary to JSObject
                call.resolve(result)
            } catch let error as SSLPinningError {
                self.reject(call, error: error)
            } catch {
                call.reject(
                    error.localizedDescription,
                    "INIT_FAILED"
                )
            }
        }
    }

    // MARK: - SSL Pinning (multiple fingerprints)

    /**
     Validates the SSL certificate of a HTTPS endpoint
     using multiple allowed fingerprints.
     */
    @objc func checkCertificates(_ call: CAPPluginCall) {
        let url = call.getString("url", "")

        // Extraction and filtering of optional fingerprints
        let fingerprints =
            call.getArray("fingerprints")?
            .compactMap { $0 as? String }
            .filter { !$0.isEmpty }

        guard !url.isEmpty else {
            call.reject(
                SSLPinningErrorMessages.urlRequired,
                "INVALID_INPUT"
            )
            return
        }

        Task {
            do {
                let result =
                    try await implementation.checkCertificates(
                        urlString: url,
                        fingerprintsFromArgs:
                            fingerprints?.isEmpty == false
                            ? fingerprints
                            : nil
                    )

                call.resolve(result)
            } catch let error as SSLPinningError {
                self.reject(call, error: error)
            } catch {
                call.reject(
                    error.localizedDescription,
                    "INIT_FAILED"
                )
            }
        }
    }

    // MARK: - Version

    /// Retrieves the plugin version synchronized from package.json.
    @objc func getPluginVersion(_ call: CAPPluginCall) {
        // Standardized enum name across all CapKit plugins
        call.resolve([
            "version": PluginVersion.number
        ])
    }

}
